### ISSUE TITLE
RFC: Improved syntax for aggregates.

### ISSUE THREAD

**Issue by ryzhyk** [vmware/differential-datalog#774]
@blp , @mbudiu-vmw 

The current syntax for aggregation is:

```
var <result variable> = Aggregate((<group-by vars>), <reduction function>(<projection expression>))
```

This has several issues:
1. the left-hand-side of the aggregation clause must have exactly one variable, e.g., this is illegal:
   ```
   (var x, var y) = Aggregate(....)
   ```

1. There is currently no way to use a lambda expression instead of `<reduction function>`.

1. Overall, the syntax is non-intuitive.

We could fix the first issue without touching anything else, but the other two require rethinking the syntax and semantics of aggregation, which is proving tricky. Here is one option that looks appealing, but does not quite work:

```
GroupBy (<group-by variables>)
```

This construct groups records in the input collection that share common values of the specified variables and returns an object of type `Group<K,V>`.  It is then up to the user what to do with this object.  They can apply any of the predefined reductions, e.g., `group_to_set`, `group_count`, etc., or build their own (groups are iterable objects that can be mapped, folded, etc.). I like this syntax as it replaces the complex `Aggregate` construct with a much more intuitive `GroupBy` and it relies on existing language features to reduce the resulting group, since groups are just another container type in in DDlog.

The problem with this is that the value type (`V`) is not well-defined.  Consider an example:

```
R(..) :- 
  R1(a: string, b: usize),
  R2(b,c),
  var g: Group<string, ???> = GroupBy(a),
  // Compute the sum of all b's in each group.
  var res = g.sum(|x| x.b).
```

Since we aggregate by `a`, the key type is `string`, hence `g` has type `Group<string, ???>`.  But what does `???` stand for? Intuitively, we want it to be a struct that represents the record with all variables available in the current syntactic scope, i.e., `a`, `b`, and `c`.  So that we can write `x.b` in order to extract the value of variable `b` for each record in the last line of the rule. Problem is, this type is not explicitly defined everywhere. We could create it implicitly, which would work in this case, but it wouldn't work for instance if we wanted to implement our own function instead of using `sum`.  We wouldn't have a way to refer to the implicitly declared type from that function.

The only solution I can think of is to require the user to explicitly construct the value type as a projection of the current record:
```
GroupBy(<group-by variables>).Project(<projection expression>)
```

where `<projection expression>` is an arbitrary expression that can refer to any of the variables visible before the `GroupBy` construct.  The the above rule can be written as:

```
R(..) :- 
  R1(a: string, b: usize),
  R2(b,c),
  var g: Group<string, usize> = GroupBy(a).Project(b),
  // Compute the sum of all b's in each group.
  var res = g.sum(|x| x).
```

or even:

```
R(..) :- 
  R1(a: string, b: usize),
  R2(b,c),
  var res: usize = GroupBy(a).Project(b).sum(|b|b).
```

One can think of `Project` as a higher-order function that takes a 0-arity closure and even make it explicit in the syntax:
```
var res: usize = GroupBy(a).project(||b).sum(|x|x).
```
where `b` is a captured variable. This way it will look similar to other container transformers like `map` and `fold`, except that it _must_ always follow `GroupBy`.

Another question is whether the argument of `GroupBy` must be a list of variables of whether we can allow arbitrary expressions there.  The former syntax has a dual purpose: it specifies the function used to extract the key from each record for use in grouping and also tells the compiler that the group-by variables must be visible after the aggregation clause; however if the expression to group by is not just a combination of existing variables, the use must introduce a new variable in a separate clause.  Example:

```
// We would like to group records that have the same sum of a and b.  Introduce a new variable for this.
var x = a+b,
var res = GroupBy(x).project(b).sum(|b|b)
```

The latter syntax eliminates the need for new variables; however it requires the user to explicitly compute and bind all values that must be visible after the `GroupBy` clause, e,g.:

```
(var x, var res) = GroupBy(a+b).project(b).(|g: Group<usize, usize>| (g.key(), g.sum(|b|b)).
```

It's hard to get excited about this more verbose syntax. But it does support arbitrary group-by expressions and it also reduces the amount of compiler "magic" by giving the user complete control over what variables will be visible after `GroupBy`, specifically only the ones bound in the left-hand-side of the `GroupBy` clause, in this case `x` and `res`.

Summary of open questions
--------------------

1. Does the `GroupBy(...).project(...)` syntax make sense?  Can we do better?
1. Should `GroupBy` take a list of variables or an arbitrary expression?

**Comment by ryzhyk** [comment]
@blp proposed the following alternative syntax that seems more intuitive than the above proposal: instead of 

```
GroupBy(<group-by variables>).Project(<projection expression>)
```
we write
```
<project expression>.group_by(<group-by variables>)
```
For instance:
```
var res: usize = (a,b).group_by(a).sum(|x|x).
```

**Comment by Kixiron** [comment]
Does `GroupBy` have the same effect of clearing the variable scope as `Aggregate` or would it allow usage like this?
```prolog
relation Mother(name: istring)
relation Children(mother: istring, name: istring)

AllChildren(mother: istring, children: Vec<istring>)
AllChildren(mother, children) :-
    Mother(mother),
    Children(mother, child),
    var children = GroupBy(child).
```
(Unless there's a way to do this currently that I'm missing)

**Comment by ryzhyk** [comment]
I think, what you mean is

```
var children = child.group_by(mother),
``` 

This will hide the `child` variable (because what could it possibly mean after aggregation?), but `mother` is still visible.

The type of `children` in this case is `Group<istring, istting>`, but it can be converted to a vector like this:

```
var children: Vec<istring> = child.group_by(mother).to_vec(),
```

**Comment by ryzhyk** [comment]
Closed via #780
